frontend/src/pages/class/[classId]/session/[sessionId]/join.tsx

```diff
@@ -91,52 +174,74 @@ const JoinSession = () => {
          * for simplicity, we will use the encryption of the student's name + email under the teacher's long term private key as the pseudonym
          */
 
+        let intervalId: NodeJS.Timeout | null = null;
+
         /**
          * wait for a confirmation from the teacher that we are allowed to join the session
          * the teacher responds with a authentication token (encrypted with the shared secret)
          * this token is then used to authenticate the student against the server
          */
+        websocketContext.socket.once(
+          "studentAuthenticationToken",
+          async (data) => {
+            // stop sending requests to the teacher
+            if (intervalId) {
+              clearInterval(intervalId);
+            }
+
+            // verify the shared secret by decrypting the confirmation message from the teacher
+            // if the message can be decrypted, store the shared secret + the authentication token
+            const authenticationToken = await ephemeralKey.decryptString(
+              decodeBase64(data.authenticationToken),
+            );
+
+            // if the message cannot be decrypted, the student is not allowed to join the session as someone is trying to impersonate the teacher
+            updateAuthenticationContext({
+              ...authenticationContext,
+              keyPair,
+              authenticationToken,
+              sessionId: sessionId,
+              ephemeralKey,
+            });
+
+            isAuthenticating.current = false;
+          },
+        );
 
-        // verify the shared secret by decrypting the confirmation message from the teacher
+        const sendRequest = async () => {
+          websocketContext.socket.emit("requestTeacherToSignInStudent", {
+            teacherId,
+            studentPublicKey: JSON.stringify(await keyPair.exportPublicKey()),
+            encryptedAuthenticationRequest: encodeBase64(
+              await ephemeralKey.encryptString(
+                JSON.stringify({
+                  classId,
+                  idToken: authenticationContext.idToken,
+                } as StudentAuthenticationRequestContent),
+              ),
+            ),
+          });
+        };
 
-        // if the message can be decrypted, store the shared secret + the authentication token
-        const authenticationToken = "retrieve from teacher";
+        // send the first request to the teacher
+        sendRequest();
 
-        // if the message cannot be decrypted, the student is not allowed to join the session as someone is trying to impersonate the teacher
-        updateAuthenticationContext({
-          ...authenticationContext,
-          keyPair,
-          authenticationToken,
-          sessionId: session.id,
-        });
+        // then repeatedly send the request to the teacher until we receive a response
+        intervalId = setInterval(sendRequest, 5 * 1000);
```

Comment:
Every 5 seconds might be a bit aggressive. Think a class of 30, that's 6 requests per second to the teacher machine. It might make sense however to have a longer waiting time (as fallback), with the server responding "the teacher is not online, retry later" for confirmation/further delay. The server could also notify "the teacher has come online" (i.e., wait a random number of seconds, then send request). What do you think? 🤔

Reply:
6 reqs/s does not seem like a lot to me. Even if all 30 sent their request at the same time - since this only happens once it should not hurt?
However, I agree that something like what you are describing would be cleaner I am not convinced by the additional value we get from doing this now compared to all the other open tickets?